README.md
# Docklet

https://unias.github.io/docklet

## Intro

Docklet is a cloud operating system for mini-datacener. Its goal is to
help multi-user share cluster resources effectively.  In Docklet, every
user has their own private **virtual cluster (vcluster)**, which
consists of a number of virtual Linux container nodes distributed over
the physical cluster. Each vcluster is separated from others and can be
operated like a real physical cluster. Therefore, most applications,
especially those requiring a cluster environment, can run in vcluster
seamlessly.

Users manage and use their vcluster all through web. The only client
tool needed is a modern web browser supporting HTML5, like Safari,
Firefox, or Chrome.  The integrated *jupyter notebook* provides a web
**Workspace**. In the Workspace, users can code, debug, test,
and run their programs, even visualize the outputs online.
Therefore, it is ideal for data analysis and processing.

Docklet creates virtual nodes from a base image. Admins can
pre-install development tools and frameworks according to their
interests. The users are also free to install their specific software
in their vcluster.

Docklet only need **one** public IP address. The vclusters are
configured to use private IP address range, e.g., 172.16.0.0/16,
192.168.0.0/16, 10.0.0.0/8. A proxy is setup to help
users visit their vclusters behind the firewall/gateway.

The Docklet system runtime consists of four components:

- distributed file system server
- etcd server
- docklet supermaster, master
- docklet worker

## Architecture
![](./docklet-arch.png)

For detailed information about configurations, please see [Config](#config).

## Install

Currently the Docklet system is recommend to run in Ubuntu 15.10+.

Ensure that python3.5 is the default python3 version.

Clone Docklet from github

```
git clone  https://github.com/unias/docklet.git
```

Run **prepare.sh** from console to install depended packages and
generate necessary configurations.

A *root* users will be created for managing the Docklet system. The
password is recorded in `FS_PREFIX/local/generated_password.txt` .

## Config ##

The main configuration file of docklet is conf/docklet.conf. Most
default setting works for a single host environment.

First copy docklet.conf.template to get docklet.conf.

Pay attention to the following settings:

- NETWORK_DEVICE : the network interface to use.
- ETCD : the etcd server address. For distributed multi hosts
  environment, it should be one of the ETCD public server address.
  For single host environment, the default value should be OK.
- STORAGE : using disk or file to storage persistent data, for
  single host, file is convenient.
- FS_PREFIX: the working dir of docklet runtime. default is
  /opt/docklet.
- CLUSTER_NET: the vcluster network ip address range, default is
  172.16.0.1/16. This network range should all be allocated to  and
  managed by docklet.
- PROXY_PORT : the listening port of configurable-http-proxy. It proxy
  connections from external public network to internal private
  container networks.
- PORTAL_URL : the portal of the system. Users access the system
  by visiting this address. If the system is behind a firewall, then
  a reverse proxy should be setup. Default is MASTER_IP:NGINX_PORT.
- NGINX_PORT : the access port of the public portal. User use this
  port to visit docklet system.
- DISTRIBUTED_GATEWAY : whether the users' gateways are distributed
  or not. Both master and worker must be set by same value.
- PUBLIC_IP : public ip of this machine. If DISTRIBUTED_GATEWAY is True,
  users' gateways can be setup on this machine. Users can visit this
  machine by the public ip. default: IP of NETWORK_DEVICE.
- USER_IP : the ip of user server. default : localhost
- MASTER_IPS : tell the web server the ips of all the cluster master.
- AUTH_KEY: the key to request users server from master, or to request
  master from users server. Please set the same value on each machine.
  Please don't use the default value.

## Start ##

### distributed file system ###

For multi hosts distributed environment, a distributed file system is
needed to store global data. Currently, glusterfs has been tested.
Lets presume the file system server export filesystem as nfs
**fileserver:/pub** :

In each physical host to run docklet, mount **fileserver:/pub** to
**FS_PEFIX/global** .

For single host environment, nothing to do.

### etcd ###

For single host environment, start **tools/etcd-one-node.sh** . Some recent
Ubuntu releases have included **etcd** in the repository, just `apt-get
install etcd`, and it need not to start etcd manually. For others, you
should install etcd manually.

For multi hosts distributed environment, **must** start
**tools/etcd-multi-nodes.sh** in each etcd server hosts. This scripts
requires users providing the etcd server address as parameters.

### supermaster ###

Supermaster is a server consist of web server, user server and a master server instance.

If it is the first time you start docklet, run `bin/docklet-supermaster init`
to init and start a docklet master, web server and user server. Otherwise, run `bin/docklet-supermaster start`.
When you start a supermaster, you don't need to start an extra master in the same cluster.

### master ###

A master manages all the workers in one data center. Docklet can manage
several data centers, each data center has one master server. But
a docklet system will only have one supermaster.

First, select a server with 2 network interface card, one having a
public IP address/url, e.g., docklet.info; the other having a private IP
address, e.g., 172.16.0.1. This server will be the master.

If it is the first time you start docklet, run `bin/docklet-master init`
to init and start docklet master. Otherwise, run  `bin/docklet-master start`,
which will start master in recovery mode in background using
conf/docklet.conf. (Note: if docklet will run in the distributed gateway mode
and recovery mode, please start the workers first.)

Please fill the USER_IP and USER_PORT in conf/docklet.conf, it is the ip and port of user server.
By default, it is `localhost` and `9100`

You can check the daemon status by running `bin/docklet-master status`

The master logs are in **FS_PREFIX/local/log/docklet-master.log** and
**docklet-web.log**.

### worker ###

Worker needs a basefs image to create containers.

You can create such an image with `lxc-create -n test -t download`,
then copy the rootfs to **FS_PREFIX/local**, and rename `rootfs`
to `basefs`.

Note the `jupyerhub` package must be installed for this image.  And the
start script `tools/start_jupyter.sh` should be placed at
`basefs/home/jupyter`.

You can check and run `tools/update-basefs.sh` to update basefs.

Run `bin/docklet-worker start`, will start worker in background.

You can check the daemon status by running `bin/docklet-worker status`.

The log is in **FS_PREFIX/local/log/docklet-worker.log**.

Currently, the worker must be run after the master has been started.

## Usage ##

Open a browser, visiting the address specified by PORTAL_URL ,
e.g., ` http://docklet.info/ `

That is it.

# Contribute #

Contributions are welcome. Please check [devguide](doc/devguide/devguide.md)


CHANGES
v0.3.2,  Dec 11, 2017
--------------------

**Bug Fix**
  * Fix the problem that some monitoring data are used before initialnizing.
  * Add some error message when start service failed.
  * Add npm registry.

**Improvement**
  * [#277] Support egress and ingress qos rate limiting.
  * [#277] Support network and ports mappings billings.
  * Support network monitoring.
  * Limit the number of users' vnodes by ip addresses.
  * Add billing detail and billing history detail
  * Replace lxc-info with lxc.Container.get_cgroup_item()

v0.3.0,  Sep 29, 2017
--------------------

**Bug Fix**
  * [#180] generated_password file no exist after master init
  * Release ip when create container failed.

**Improvement**
  * [#16] display file size, modification time in jupyter notebook 
  * [#86] time display in UserList
  * [#87] add a new panel to approve or decline user activation requests
  * [#121] Autofilling may lead to a bug that makes local users cannot login
  * [#178] record and display history of all containers
  * [#210] rename Dashboard to Workspace
  * [#212] add docklet hyperlink in web portal 
  * Separate user module from master.
  * Support multiple masters run in the same time, and the user can choose which to use in webpage.
  * Support distributed gateway, if enable, worker will setup its own gateway.
  * Support user gateway.

v0.2.8,  Jul 28, 2016
--------------------

**Bug Fix**
  * [#119] version display error

**Improvement**
  * [#52] give user a total quota, let themselves decide how to use quota
  * [#72] recording the user's historical resource usage
  * [#85] Making workers's state consistent with master
  * [#88] setting config file in admin panel
  * [#96] Web notifications
  * [#113] Recovery : after poweroff, just recover container, not recover service

v0.2.7,  May 17, 2016
--------------------

**Bug Fix**
  * [#9] updating user profile taking effects immediately
  * [#12] logging user's activity
  * [#14] Can't stop vcluster by dashboard page
  * [#18] subprocess call should check return status
  * [#19] lxc config string in config file is limited in 16 bytes
  * [#25] bug of external login
  * [#30] support lxc.custom.conf in appending
  * [#35] nfs mountpoint bug in imagemgr.py
  * [#49] Fail to create container
  * [#57] status page of normal user failed
  * [#68] Not Found error when just click "Sign in" Button
  * [#76] unable to show and edit user table in smartphone

**Improvement**
  * [#7] enhance quota management
  * [#8] independent starting of master and workers
  * [#20] check typing and input on web pages and web server
  * [#23] add LXCFS for container
  * [#41] move system data to global/sys
  * [#42] check IP and network pool when releasing IPs
  * [#48] token expires after some time
  * [#54] display container owner
  * [#61] rewrite httprest.py using flask routing

**Notes**
  * If you upgrade from former version, please run tools/upgrade.py first.

v0.2.6,  Mar 31, 2016
--------------------

An initial release on github.com

* Using the open source AdminLTE theme


src/env.py
import os,netifaces

def getenv(key):
    if key == "CLUSTER_NAME":
        return os.environ.get("CLUSTER_NAME", "docklet-vc")
    elif key == "FS_PREFIX":
        return os.environ.get("FS_PREFIX", "/opt/docklet")
    elif key == "CLUSTER_SIZE":
        return int(os.environ.get("CLUSTER_SIZE", 1))
    elif key == "CLUSTER_NET":
        return os.environ.get("CLUSTER_NET", "172.16.0.1/16")
    elif key == "CONTAINER_CPU":
        return int(os.environ.get("CONTAINER_CPU", 100000))
    elif key == "CONTAINER_DISK":
        return int(os.environ.get("CONTAINER_DISK", 1000))
    elif key == "CONTAINER_MEMORY":
        return int(os.environ.get("CONTAINER_MEMORY", 1000))
    elif key == "DISKPOOL_SIZE":
        return int(os.environ.get("DISKPOOL_SIZE", 10000))
    elif key == "ETCD":
        return os.environ.get("ETCD", "localhost:2379")
    elif key == "NETWORK_DEVICE":
        return os.environ.get("NETWORK_DEVICE", "eth0")
    elif key == "MASTER_IP":
        return os.environ.get("MASTER_IP", "0.0.0.0")
    elif key == "MASTER_IPS":
        return os.environ.get("MASTER_IPS", "0.0.0.0@docklet")
    elif key == "MASTER_PORT":
        return int(os.environ.get("MASTER_PORT", 9000))
    elif key == "WORKER_PORT":
        return int(os.environ.get("WORKER_PORT", 9001))
    elif key == "NGINX_PORT":
        return int(os.environ.get("NGINX_PORT", 8080))
    elif key == "PROXY_PORT":
        return int(os.environ.get("PROXY_PORT", 8000))
    elif key == "PROXY_API_PORT":
        return int(os.environ.get("PROXY_API_PORT", 8001))
    elif key == "WEB_PORT":
        return int(os.environ.get("WEB_PORT", 8888))
    elif key == "PORTAL_URL":
        return os.environ.get("PORTAL_URL",
            "http://"+getenv("MASTER_IP") + ":" + str(getenv("NGINX_PORT")))
    elif key == "LOG_LEVEL":
        return os.environ.get("LOG_LEVEL", "DEBUG")
    elif key == "LOG_LIFE":
        return int(os.environ.get("LOG_LIFE", 10))
    elif key == "WEB_LOG_LEVEL":
        return os.environ.get("WEB_LOG_LEVEL", "DEBUG")
    elif key == "STORAGE":
        return os.environ.get("STORAGE", "file")
    elif key =="EXTERNAL_LOGIN":
        return os.environ.get("EXTERNAL_LOGIN", "False")
    elif key =="DATA_QUOTA":
        return os.environ.get("DATA_QUOTA", "False")
    elif key =="DATA_QUOTA_CMD":
        return os.environ.get("DATA_QUOTA_CMD", "gluster volume quota docklet-volume limit-usage %s %s")
    elif key == 'DISTRIBUTED_GATEWAY':
        return os.environ.get("DISTRIBUTED_GATEWAY", "False")
    elif key == "PUBLIC_IP":
        device = os.environ.get("NETWORK_DEVICE","eth0")
        addr = netifaces.ifaddresses(device)
        if 2 in addr:
            return os.environ.get("PUBLIC_IP",addr[2][0]['addr'])
        else:
            return os.environ.get("PUBLIC_IP","0.0.0.0")
    elif key == "NGINX_CONF":
        return os.environ.get("NGINX_CONF","/etc/nginx")
    elif key =="USER_IP":
        return os.environ.get("USER_IP","0.0.0.0")
    elif key =="USER_PORT":
        return int(os.environ.get("USER_PORT",9100))
    elif key =="AUTH_KEY":
        return os.environ.get("AUTH_KEY","docklet")
    elif key =="OPEN_REGISTRY":
        return os.environ.get("OPEN_REGISTRY","False")
    elif key =="APPROVAL_RBT":
        return os.environ.get("APPROVAL_RBT","ON")
    elif key =="ALLOCATED_PORTS":
        return os.environ.get("ALLOCATED_PORTS","10000-65535")
    else:
        return os.environ.get(key,"")
